src/software/new_geom/ray.h

```diff
@@ -0,0 +1,106 @@
+#pragma once
+
+#include "software/geom/line.h"
+#include "software/geom/point.h"
+
+class Ray final
+{
+   public:
+    /**
+     * Creates a degenerate Ray at (0, 0)
+     */
+    inline explicit constexpr Ray() {}
+
+    /**
+     * Creates a Ray with a start point and a direction angle
+     *
+     * @param start the start point of Ray
+     * @param direction the direction angle of Ray
+     */
+    inline explicit Ray(const Point& start, const Angle& direction)
+        : start_(start), direction_(direction)
+    {
+    }
+
+    /**
+     * Creates a Ray with a start point and a direction vector
+     *
+     * @param start the start point of Ray
+     * @param direction the direction vector of Ray
+     */
+    inline explicit Ray(const Point& start, const Vector& direction)
+        : Ray(start, direction.orientation())
+    {
+    }
+
+    /**
+     * Returns the start point of this Ray
+     *
+     * @return the start point of this Ray
+     */
+    inline Point getStart() const
+    {
+        return start_;
+    }
+
+    /**
+     * Returns the direction of this Ray
+     *
+     * @return the direction of this Ray
+     */
+    inline Angle getDirection() const
+    {
+        return direction_;
+    }
+
+    /**
+     * Sets the start point of this Ray
+     *
+     * @param point the new start point
+     */
+    inline void setStart(const Point& start)
+    {
+        start_ = start;
+    }
+
+    /**
+     * Sets the direction of this Ray with an angle
+     *
+     * @param direction the new direction as an angle
+     */
+    inline void setDirection(const Angle& direction)
+    {
+        direction_ = direction;
+    }
+
+    /**
+     * Sets the direction of this Ray with a vector
+     *
+     * @param direction the new direction as a vector
+     */
+    inline void setDirection(const Vector& direction)
+    {
+        direction_ = direction.orientation();
+    }
+
+    /**
+     * Rotates this Ray counterclockwise by an angle
+     *
+     * @param angle the angle to rotate the Ray
+     */
+    inline void rotate(Angle angle)
+    {
+        direction_ += angle;
```

Comment:
Might make sense to do an AngleMod here so that we don't potentially go over 360 degrees.

Reply:
Really something the += operator should handle @jsn0716 can you that it does or if new_geom will?

Reply:
No, it doesn't handle that now so `angleMod` (or `clamp`, not sure why there are two functions that do the same thing) needs to be explicitly called to ensure that it's in the correct range. The `angle` class is being ported in #970, but if implicitly keeping the `Angle` within `[−π, π]` is something we want, I think we should create a separate issue.

Reply:
Agree with separate issues, and agree we may want to combine `angleMod` and `clamp`. @jsn0716 can you make an issue for that?

Reply:
Done: #988